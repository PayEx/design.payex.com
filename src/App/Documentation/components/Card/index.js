import React, { Component } from "react";
import { Link } from "react-router-dom";

import { ComponentPreview, Property, DocToc, ExperimentalComponentAlert } from "#";
import CardComponent from "@/Card";
import MediaObject from "@/MediaObject";

const DefaultBehaviour = () => (
    <>
        <h2 id="default-behaviour">Default behaviour</h2>
        <p>
            Wrap indiviual cards in the <Property value=".cards" /> class to display them lined up next to each other.
            When the window reaches <Link to="/docs/core/breakpoints">breakpoint</Link> sm the cards will switch to mobile view and take up one row each and stretch to fill the width of the container.
        </p>
        <ComponentPreview language="html" showCasePanel codeFigure>
            <div className="cards">
                <CardComponent
                    type="default"
                    title="Card default"
                    imgSrc="https://picsum.photos/300"
                    text="This is the cards text"
                    smallText="And the small text"
                    btnTxt="Button"
                    footerTxt="This is some text that goes in the footer."
                    footerLinkTxt="You wanna click this link?"
                >
                    <MediaObject size="sm" icon="account_circle" heading="Bob Corlsan" text="bob.corlsan@payex.com" textSmall muted />
                </CardComponent>

                <CardComponent
                    type="default"
                    title="Card default"
                    imgSrc="https://picsum.photos/300"
                    text="This is the cards text"
                    smallText="And the small text"
                    btnTxt="Button"
                    footerTxt="This is some text that goes in the footer."
                    footerLinkTxt="You wanna click this link?"
                >
                    <MediaObject size="sm" icon="account_balance" heading="725 NOK" text="Outstanding balance" textSmall muted />
                </CardComponent>

                <CardComponent
                    type="default"
                    title="Card default"
                    imgSrc="https://picsum.photos/300"
                    text="This is the cards text"
                    smallText="And the small text"
                    btnTxt="Button"
                    footerTxt="This is some text that goes in the footer."
                    footerLinkTxt="You wanna click this link?"
                >
                    <MediaObject size="sm" icon="verified_user" heading="75%" text="Complete" textSmall muted />
                </CardComponent>
            </div>
        </ComponentPreview>
    </>
);

const CardTypes = () => (
    <>
        <h2 id="card-types">Card Types</h2>
        <p>To decide how your cards will look use <Property value=".card-default" />, <Property value=".card-brand" />, or <Property value=".card-light" /></p>
        <ComponentPreview language="html" showCasePanel codeFigure>
            <div className="cards">
                <CardComponent
                    type="default"
                    title="Card default"
                    imgSrc="https://picsum.photos/300"
                    text="This is the cards text"
                    smallText="And the small text"
                    btnTxt="Button"
                    footerTxt="This is some text that goes in the footer."
                    footerLinkTxt="You wanna click this link?"
                >
                    <MediaObject size="sm" icon="account_circle" heading="Bob Corlsan" text="bob.corlsan@payex.com" textSmall muted />
                </CardComponent>

                <CardComponent
                    type="brand"
                    title="Card brand"
                    imgSrc="https://picsum.photos/300"
                    text="This is the cards text"
                    smallText="And the small text"
                    btnTxt="Button"
                    footerTxt="This is some text that goes in the footer."
                    footerLinkTxt="You wanna click this link?"
                >
                    <MediaObject size="sm" icon="account_balance" heading="725 NOK" text="Outstanding balance" textSmall muted />
                </CardComponent>

                <CardComponent
                    type="light"
                    title="Card light"
                    imgSrc="https://picsum.photos/300"
                    text="This is the cards text"
                    smallText="And the small text"
                    btnTxt="Button"
                    footerTxt="This is some text that goes in the footer."
                    footerLinkTxt="You wanna click this link?"
                >
                    <MediaObject size="sm" icon="verified_user" heading="75%" text="Complete" textSmall muted />
                </CardComponent>
            </div>
        </ComponentPreview>
    </>
);

const CardText = () => (
    <div className="col-md-12 col-lg-10 doc-body">
        <ExperimentalComponentAlert />
        <p className="lead">Card...</p>
        <DefaultBehaviour />
        <CardTypes />
    </div>
);

class Card extends Component {
    render () {
        return (
            <div className="doc-container">
                <div className="row">
                    <CardText />
                    <DocToc component={CardText} />
                </div>
            </div>
        );
    }
}

export default Card;

/* For testing */
export { DefaultBehaviour, CardTypes, CardText };
