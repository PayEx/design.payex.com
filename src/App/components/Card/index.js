import React from "react";
import PropTypes from "prop-types";

const Card = ({ type, title, imgSrc, text, smallText, btnTxt, footerTxt, footerLink, footerLinkTxt, children }) => (
    <div className={`card${type ? ` card-${type}` : ""}`}>
        {title ? <header>
            <h3>
                <a href="#">{title}</a>
            </h3>
        </header> : null}
        {imgSrc ? <div className="card-img">{"\n"}
            <img src={imgSrc} alt=""/>{"\n"}
        </div> : null}
        <div className="highlight">
            {children}
        </div>
        <div className="card-body">
            {text ? <p>{text}</p> : null}{"\n"}
            {smallText ? <small>{smallText}</small> : null}
        </div>
        <footer>
            {btnTxt ? <button type="button" className="btn btn-primary btn-block">{btnTxt}</button> : null}
            {footerTxt ? <div className="footer-text"><p>{footerTxt}</p></div> : null}
            {footerLinkTxt ? <div className="footer-link"><a href={footerLink || "#"}>{footerLinkTxt}</a></div> : null}
        </footer>
    </div>
);

Card.propTypes = {
    type: PropTypes.oneOf(["brand", "default", "light"]),
    title: PropTypes.string,
    imgSrc: PropTypes.string,
    icon: PropTypes.string,
    text: PropTypes.string,
    smallText: PropTypes.string,
    btnTxt: PropTypes.string,
    footerTxt: PropTypes.string,
    footerLink: PropTypes.string,
    footerLinkTxt: PropTypes.string
};

export default Card;
