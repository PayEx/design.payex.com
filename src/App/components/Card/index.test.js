import React from "react";
import { shallow } from "enzyme";

import Card from "./index";

describe("Component: Card - ", () => {
    it("is defined", () => {
        expect(Card).toBeDefined();
    });

    it("throws an error if type does not match proptypes", () => {
        console.error = jest.fn();

        const wrapper = shallow(<Card type="error" />);

        expect(console.error).toHaveBeenCalled();
        expect(wrapper).toMatchSnapshot();
    });

    it("renders with class name card", () => {
        const wrapper = shallow(<Card />);

        expect(wrapper.html()).toContain("card");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders with correct class name", () => {
        const wrapper = shallow(<Card type="brand"/>);

        expect(wrapper.html()).toContain("card-brand");
        expect(wrapper.html()).not.toContain("card-default");
        expect(wrapper.html()).not.toContain("card-light");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders a title", () => {
        const wrapper = shallow(<Card title="Card title goes here"/>);

        expect(wrapper.html()).toContain("header");
        expect(wrapper.find("header").text()).toEqual("Card title goes here");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders an image", () => {
        const wrapper = shallow(<Card imgSrc="url.address" />);

        expect(wrapper.html()).toContain("card-img");
        expect(wrapper.find("img").prop("src")).toEqual("url.address");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders a text in card-body", () => {
        const wrapper = shallow(<Card text="Card text" />);

        expect(wrapper.html()).toContain("p");
        expect(wrapper.find("p")).toHaveLength(1);
        expect(wrapper.find("p").text()).toEqual("Card text");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders a smallText in card-body", () => {
        const wrapper = shallow(<Card smallText="small text" />);

        expect(wrapper.html()).toContain("small");
        expect(wrapper.find("small")).toHaveLength(1);
        expect(wrapper.find("small").text()).toEqual("small text");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders a button in the footer", () => {
        const wrapper = shallow(<Card btnTxt="Button" />);

        expect(wrapper.html()).toContain("btn");
        expect(wrapper.find(".btn")).toHaveLength(1);
        expect(wrapper.find(".btn").text()).toEqual("Button");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders footer text", () => {
        const wrapper = shallow(<Card footerTxt="Footer information" />);

        expect(wrapper.html()).toContain("footer-text");
        expect(wrapper.find(".footer-text")).toHaveLength(1);
        expect(wrapper.find(".footer-text").text()).toEqual("Footer information");
        expect(wrapper).toMatchSnapshot();
    });

    it("renders footer link text", () => {
        const wrapper = shallow(<Card footerLinkTxt="Click this footer link" />);

        expect(wrapper.html()).toContain("footer-link");
        expect(wrapper.find(".footer-link")).toHaveLength(1);
        expect(wrapper.find(".footer-link").text()).toEqual("Click this footer link");
        expect(wrapper).toMatchSnapshot();
    });
});
