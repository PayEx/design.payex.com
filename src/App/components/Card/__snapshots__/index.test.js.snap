// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Card -  renders a button in the footer 1`] = `
<div
  className="card"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer>
    <button
      className="btn btn-primary btn-block"
      type="button"
    >
      Button
    </button>
  </footer>
</div>
`;

exports[`Component: Card -  renders a smallText in card-body 1`] = `
<div
  className="card"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

    <small>
      small text
    </small>
  </div>
  <footer />
</div>
`;

exports[`Component: Card -  renders a text in card-body 1`] = `
<div
  className="card"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    <p>
      Card text
    </p>
    

  </div>
  <footer />
</div>
`;

exports[`Component: Card -  renders a title 1`] = `
<div
  className="card"
>
  <header>
    <h3>
      <a
        href="#"
      >
        Card title goes here
      </a>
    </h3>
  </header>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer />
</div>
`;

exports[`Component: Card -  renders an image 1`] = `
<div
  className="card"
>
  <div
    className="card-img"
  >
    

    <img
      alt=""
      src="url.address"
    />
    

  </div>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer />
</div>
`;

exports[`Component: Card -  renders footer link text 1`] = `
<div
  className="card"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer>
    <div
      className="footer-link"
    >
      <a
        href="#"
      >
        Click this footer link
      </a>
    </div>
  </footer>
</div>
`;

exports[`Component: Card -  renders footer text 1`] = `
<div
  className="card"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer>
    <div
      className="footer-text"
    >
      <p>
        Footer information
      </p>
    </div>
  </footer>
</div>
`;

exports[`Component: Card -  renders with class name card 1`] = `
<div
  className="card"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer />
</div>
`;

exports[`Component: Card -  renders with correct class name 1`] = `
<div
  className="card card-brand"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer />
</div>
`;

exports[`Component: Card -  throws an error if type does not match proptypes 1`] = `
<div
  className="card card-error"
>
  <div
    className="highlight"
  />
  <div
    className="card-body"
  >
    

  </div>
  <footer />
</div>
`;
